### PR TITLE
Implement unicode property escape matching

### PR DESCRIPTION
**Pull request by raskad**
This is my attempt at getting unicode property escapes working.

Info on the feature:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes
- https://tc39.es/ecma262/#sec-patterns

The tables for the implemented unicode properties: https://tc39.es/ecma262/#table-nonbinary-unicode-properties

There are some open points:

1. Property escapes in brackets
Unicode property escapes may be used in brackets e.g. `[\p{ASCII}]`. I have not implemented this currently. Based on my understanding `BracketContents` and `CodePointSet` would have to undergo some heavy refactoring for this to work.
There is a TODO marked at `Insn::Bracket(BracketContents)` that reads `This is a very heavyweight instruction, consider breaking it up.` This may be a good opportunity to do a refactor here, but I am not 100% sure how to go about that.
2. Tests
I have pulled all of the invalid test cases from the 262 suite into `property_escapes_invalid`. The 262 suite contains generated, exhaustive tests for all properties. I wrote an example port of that type of test for one script property `unicode_escape_property_script_buhid`. Generating those type of test would work, but we would have to run those tests in release mode, because they are exhaustive and test every unicode codepoint per property. This takes way to long in debug mode. I am not really convinced, that it makes much sense to generate those tests, as we generate the unicode tables in the same step. But currently have no good ideas about better testing methods.
3. Code generation
I have used the `codegen` crate in `gen-unicode`, as there is a lot more generation going on now. I personally think it is easier to read than the raw string generation and has potential for some better code deduplication. But that is only my opinion.

### CODE REVIEW

**Comment by raskad** [comment]
@ridiculousfish I would really appreciate your input on the PR and the three points in the PR description.

**Review by ridiculousfish — CHANGES_REQUESTED** [review]
This is awesome!! Very impressive, I'm very pleased with the overall approach. Pulling in the `codegen` crate is fine since this only affects `gen_unicode` and not the main library.

A minor concern is the implementation of `insn::Insn::UnicodePropertyEscape` which ends up calling `is_character_class`. A more efficient implementation would be to store a slice of `CodePointRange`, and binary search that, instead of going through a branch in `is_character_class`. Maybe a separate instruction for `CodePointRangeUnpacked`. But this is a minor optimization and you can implement it now, later, or not at all, as you please.

A major concern is that the tests are incomplete. There's great tests for parse failures, but the positive tests are all focused on Buhid. For example I can make `is_property_value_general_category` just always return false, and the tests still pass. Are there more tests from test262 or wherever which we can pull in?

File: src/classicalbacktrack.rs [hunk 1]
```diff
@@ -849,6 +849,21 @@ impl<'a, Input: InputIndexer> MatchAttempter<'a, Input> {
                             break 'backtrack;
                         }
                     }
+
+                    Insn::UnicodePropertyEscape {
+                        property_escape,
+                        negate,
+                    } => {
+                        let m = if (scm::UnicodePropertyEscape { property_escape })
```
[the remark this answers is not in the document]
Comment (raskad) [reply src/classicalbacktrack.rs]: Implemented in c66e662

File: src/unicode.rs [hunk 2]
```diff
@@ -269,3 +292,70 @@ pub(crate) fn is_id_continue(c: char) -> bool {
     let i = c as u32;
     ID_CONTINUE.binary_search_by(|&cpr| cpr.compare(i)).is_ok()
 }
+
+#[derive(Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (raskad) [reply src/unicode.rs]: Completely right. They should be `Copy`.
Implemented in c66e662

File: tests/unicode_escapes.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        (0x00E000, 0x10FFFF),
+    ];
+    let regexes = vec![
+        r#"^\P{Script=Buhid}+$"#,
+        r#"^\P{Script=Buhd}+$"#,
+        r#"^\P{sc=Buhid}+$"#,
+        r#"^\P{sc=Buhd}+$"#,
+    ];
+
+    run_property_escape_test(tc, regexes, build_test_string(lone_code_points, ranges));
```
[the remark this answers is not in the document]
Comment (raskad) [reply tests/unicode_escapes.rs]: I have implemented some generated tests in 5946597 that only check some code points. They run much faster. And I have uncovered some bugs in my codegen so they are already working :)

There are still some tests missing for some derived properties and the special cases.

**Comment by raskad** [comment]
@ridiculousfish I have implemented some further packing(combining) of code points ranges in d528358. The logic seems fine to me and the tests pass, but maybe you can review that commit separately ;)

**Comment by raskad** [comment]
@ridiculousfish I have added tests the rest of the tests. I think we should push the implementation of unicode property escapes in brackets to a different PR, as there is already a lot happening here.

**Review by ridiculousfish — APPROVED** [review]
This looks excellent! Amazing contribution, thank you!

**Pull request merged**